### PR TITLE
refactor(framework): Rename inputs to controls

### PR DESCRIPTION
**Pull request by rifont**
### What changed? Why was the change needed?
This is the first of a sequence of 3 PRs to replace the "inputs" nomenclature with "controls" in the Novu Framework, API, and Worker:
1. (THIS PR): In `@novu/framework`:
  * Rename `inputs` -> `controls` in client API callbacks (with `@deprecated` flag on `inputs`)
  * Rename `inputSchema` -> `controlSchema` in client Workflow/Step options (with `@deprecated` flag on `inputSchema`)
  * Accept both `inputs` and `controls` during Execution
  * Return both `inputSchema` and `controlSchema` during Discovery
2. Next PR: Update API/Worker to:
  * Consume the `controlSchema` during Discovery
  * Provide `controls` during Execution
3. Sunset PR: remove all references to `inputs` and `inputSchema` across both Framework and API

### Screenshots

_Testing backwards compatibility_
Using a new `@novu/framework` SDK with current Novu Prod API/Worker
IMG_7964

<details>
<summary><strong>Expand for optional sections</strong></summary>

### Special notes for your reviewer

</details>

### CODE REVIEW

File: packages/framework/src/client.test.ts [hunk 1]
```diff
@@ -215,18 +215,18 @@ describe('Novu Client', () => {
           }),
           {
             providers: {
-              slack: async ({ inputs }) => {
-                const blocks = [
-                  {
-                    type: 'header' as any,
```
Comment (rifont) [inline packages/framework/src/client.test.ts:255]: Small refactor to place `blocks` directly on the provider return params to remove the `as any` on `type`

**Comment by scopsy** [comment]
Did you modified the Workflow Step editor page (new designs) there is a tab there called "Step Inputs"

File: packages/framework/src/client.test.ts [hunk 2]
```diff
@@ -299,6 +299,7 @@ describe('Novu Client', () => {
         },
         state: [],
         inputs: {},
+        controls: {},
```
Comment (rifont) [inline packages/framework/src/client.test.ts:336]: In the #3 sunset PR, we will remove `inputs` from here.

File: packages/framework/src/client.test.ts [hunk 3]
```diff
@@ -469,9 +473,49 @@ describe('Novu Client', () => {
...
+        data: {},
+        inputs: {},
+        controls: {
+          foo: 'foo',
+        },
+      };
+
+      const executionResult = await client.executeWorkflow(event);
+
+      expect(executionResult.providers).toEqual({ sendgrid: { ipPoolName: 'foo Subject' } });
```
Comment (rifont) [inline packages/framework/src/client.test.ts:552]: We previously weren't providing the step outputs to the Provider resolution. This has been fixed, and this test is added to verify the correct behaviour.

File: packages/framework/src/client.ts [hunk 4]
```diff
@@ -4,10 +4,10 @@ import ora from 'ora';
 import { HttpHeaderKeysEnum } from './constants';
 import {
   ExecutionEventDataInvalidError,
-  ExecutionEventInputInvalidError,
+  ExecutionEventControlsInvalidError,
   ExecutionProviderOutputInvalidError,
   ExecutionStateCorruptError,
-  ExecutionStateInputInvalidError,
+  ExecutionStateControlsInvalidError,
```
Comment (rifont) [inline packages/framework/src/client.ts:10]: Renaming the Errors to align with the new naming.

File: packages/framework/src/client.ts [hunk 5]
```diff
@@ -445,14 +446,18 @@ export class Client {
     console.log(`  └ ${EMOJI.DURATION} duration: '${duration.toFixed(2)}ms'\n`);
   }
 
-  private async executeProviders(payload: IEvent, step: DiscoverStepOutput): Promise<Record<string, unknown>> {
+  private async executeProviders(
+    payload: IEvent,
+    step: DiscoverStepOutput,
+    outputs: Record<string, unknown>
```
Comment (rifont) [inline packages/framework/src/client.ts:452]: This is the fix to ensure the outputs of the step are provided to the `providers` for re-use. A good example use-case is re-using the `chat.body` in the Slack blocks.

File: packages/framework/src/client.ts [hunk 6]
```diff
@@ -518,24 +530,24 @@ export class Client {
...
-        const templateInputs = await this.createStepInputs(step, event);
-        const inputs = await this.compileInputs(templateInputs, event);
-        const result = await step.resolve(inputs);
-        const validatedResult = await this.validate(
-          result,
+        const templateControls = await this.createStepControls(step, event);
+        const controls = await this.compileControls(templateControls, event);
+        const output = await step.resolve(controls);
+        const validatedOutput = await this.validate(
+          output,
```
Comment (rifont) [inline packages/framework/src/client.ts:537]: Some renaming in places like this to reflect the correct/updated names.

File: packages/framework/src/client.ts [hunk 7]
```diff
@@ -518,24 +530,24 @@ export class Client {
...
           step.outputs.unknownSchema,
           'step',
           'output',
           event.workflowId,
           step.stepId
         );
 
-        const providers = await this.executeProviders(event, step);
+        const providers = await this.executeProviders(event, step, validatedOutput);
```
Comment (rifont) [inline packages/framework/src/client.ts:545]: Part of the fix to ensure providers get the step output.

File: packages/framework/src/client.validation.test.ts [hunk 8]
```diff
@@ -17,37 +17,42 @@ describe('validation', () => {
...
-                inputs.foo = 123;
+                controls.foo = 123;
                 // @ts-expect-error - Type 'string' is not assignable to type 'number'.
-                inputs.baz = '123';
+                controls.baz = '123';
+
+                // @ts-expect-error - Type 'number' is not assignable to type 'string'.
+                outputs.body = 123;
+                // @ts-expect-error - Type 'number' is not assignable to type 'string'.
+                outputs.subject = 123;
```
Comment (rifont) [inline packages/framework/src/client.validation.test.ts:55]: Additional compilation test case to make sure that the `providers` get the typed `outputs`.

File: packages/framework/src/types/step.types.ts [hunk 9]
```diff
@@ -52,37 +53,62 @@ type StepOutput<T_Result> = Promise<T_Result & StepContext>;
...
+    /**
+     * The schema for the controls of the step. Used to validate the user-provided controls from Novu Web.
+     *
+     * @deprecated Use `controlSchema` instead
+     */
+    inputSchema?: T_ControlSchema;
+    /**
+     * The schema for the controls of the step. Used to validate the user-provided controls from Novu Web.
+     */
+    controlSchema?: T_ControlSchema;
```
Comment (rifont) [inline packages/framework/src/types/step.types.ts:93]: Improved type-docs.

**Comment by scopsy** [comment]
And another things, we have a few places in the UI that uses the stored control schema to render react json schema form on web. I believe we need to update it there aswell with a fallback to inputs

**Review by scopsy — APPROVED** [review]

**Review by SokratisVidros — APPROVED** [review]

**Pull request merged**